### DISCUSSION TITLE
How can I re-render canvas without erasing custom ctx additions?

### DISCUSSION THREAD

**Discussion by ghost** [fabricjs/fabric.js#9598] [Q&A] [answered]
I'm working on a canvas workspace where you do your work on the client side and send it to BE to render it into a full image.
As part of this work, I have implemented an addition to `IText` where you can do background to the text itself. The implementation is not done on fabric but directly on the canvas context. here an example for what I do:

```
ctx.moveTo(leftX, lineBottomY - maxCurveLength * roundness)
ctx.lineTo(leftX, lineTopY + maxCurveLength * roundness)
ctx.quadraticCurveTo(leftX, lineTopY, leftX + maxCurveLength * roundness, lineTopY)

// top side + right top corner
ctx.lineTo(rightX - maxCurveLength * roundness, lineTopY)
ctx.quadraticCurveTo(rightX, lineTopY, rightX, lineTopY + maxCurveLength * roundness)
// right side
ctx.lineTo(rightX, lineBottomY - maxCurveLength * roundness)
ctx.stroke()
ctx.fill()
ctx.restore()
```

However, the rendering is being done **on top of the actual text**, and I need to do `text.bringForward()` to overcome it. It also requires me to trigger `canvas.renderAll()` to make the changes visible, but it clears up all the custom ctx additions I did. 

Attached image showing the output:
<img width="936" alt="image" src="https://github.com/fabricjs/fabric.js/assets/121938774/d566ca8d-af9e-4b31-bb63-7697543f06fc">

When I use the bringForward+renderAll, this is what happening:
<img width="963" alt="image" src="https://github.com/fabricjs/fabric.js/assets/121938774/0e282ac9-89ef-4bf5-a858-eb127adcecdf">

I understand the `renderAll` function completely cleans the canvas and re-renders only the registered objects. But I wondered if there is a way to re-render without wiping the ctx custom additions or if my whole approach for implementing this feature is wrong.

Thanks!

**Comment by ShaMan123** [comment, marked as answer]
You are impl it in a wrong approach.
This is a text background, right?
So the text class rendering lifecycle is in charge of it.
Ovverride the relevant rendering methods of text such as `_render`

**Reply by ghost** [reply]
**EDIT:** seems like the function `render` (not the underscore one) works on node.

while it working nicely on frontend (thanks!) it seems like when trying implement something similar on `node-canvas` fails.
```
      const canvas = new fabric.StaticCanvas(null, {
        width: targetWidth,
        height: targetHeight,
      })

      const ctx = canvas.getContext()
      ctx.textDrawingMode = 'glyph'

      await canvas.loadFromJSON(json, () => {
        canvas.getObjects().forEach((obj: fabric.Object) => {
          // if objects text, render text background
          if (obj.type === 'textbox') {
            // renderTextBackgroundColor(obj as fabric.Textbox, canvas, canvas.getContext())
            obj._render = (ctx: any) => {
              console.log('do something')
            }
          }
        })
       ....

```
the `_render` is not being called at all. not sure if my implementation is correct, but I'm not sure how to override it after loading from a json.

**Reply by ShaMan123** [reply]
weird.
it should work.
but the official way to do this is to use `classRegistry` before you load from json ofcourse
```
classRegistry.setClass(MyText)
```
